addons/product/models/product.py

```diff
@@ -143,9 +143,13 @@ class ProductProduct(models.Model):
         'product.packaging', 'product_id', 'Product Packages',
         help="Gives the different ways to package the same product.")
 
-    _sql_constraints = [
-        ('barcode_uniq', 'unique(barcode)', "A barcode can only be assigned to one product !"),
-    ]
+    @api.one
+    @api.constrains('barcode')
+    def _check_unique_barcode_company_wise(self):
+        similar_barcode = self.search([('barcode', '=', self.barcode), ('product_tmpl_id', '!=', self.product_tmpl_id.id), ('company_id', '=', self.company_id.id)])
+        if similar_barcode:
```

Comment:
As most products are defined without a company, it is like the barcode is shared between.  I would suggest to put '|', ('company_id', '=', False), ('company_id', '=', self.company_id.id)]  It might be better to avoid that a product without and with company have the same barcode.

Reply:
<code>(['company_id', 'in', [False, self.company_id.id])</code> is shorter :)

Reply:
Need to put domain as we can add global product.